### ISSUE TITLE
TBI - `@sap-ux/project-access` update `getCapModelAndServices` to return the correct service url path format for cds version

### ISSUE THREAD

**Issue by IainSAP** [SAP/open-ux-tools#1796]
### Description (include screenshots)
CDS 6 -> 7 introduced a breaking change of service path format. We are not handling this in `getCapModelAndServices` but we we already detect the cds version used by the project and so should modify the service path to return the corrected path. This could be done optionally to allow consumers time to remove their handling of the different paths.

Propsal :

```
export async function getCapModelAndServices(
    projectRoot: string | { projectRoot: string; logger?: Logger }, updateCds7Paths?: boolean
): Promise<{ model: csn; services: ServiceInfo[] }> {
    let _projectRoot;
    let _logger;
    if (typeof projectRoot === 'object') {
        _projectRoot = projectRoot.projectRoot;
        _logger = projectRoot.logger;
    } else {
        _projectRoot = projectRoot;
    }

    const cds = await loadCdsModuleFromProject(_projectRoot);
    const capProjectPaths = await getCapCustomPaths(_projectRoot);
    const modelPaths = [
        join(_projectRoot, capProjectPaths.app),
        join(_projectRoot, capProjectPaths.srv),
        join(_projectRoot, capProjectPaths.db)
    ];
    const model = await cds.load(modelPaths, { root: _projectRoot });

    _logger?.info(`@sap-ux/project-access:getCapModelAndServices - Using 'cds.home': ${cds.home}`);
    _logger?.info(`@sap-ux/project-access:getCapModelAndServices - Using 'cds.version': ${cds.version}`);
    _logger?.info(`@sap-ux/project-access:getCapModelAndServices - Using 'cds.root': ${cds.root}`);

    let services = cds.compile.to.serviceinfo(model, { root: _projectRoot }) ?? [];
    // Update to modify the path to be correct for the cds version
    const isPreCDS7 =  lt(cds.version, '7.0.0');
    if (services.map) {
        services = services.map((value) => {
            return {
                name: value.name,
                urlPath: updateCds7Paths && isPreCDS7 ? updateToLegacyCdsServicePath(value.urlPath) : uniformUrl(value.urlPath),
                runtime: value.runtime
            };
        });
    }
    return {
        model,
        services
    };
}

/**
 * Updates the service paths to the pre CDS 7 service path format.
 * 
 */
function updateToLegacyCdsServicePath(capServicePath: string): string {
    return uniformUrl(capServicePath).replace('odata/v4/', '');
}
```

### Value
Optional - details on how this technical backlog item provides value to end-users or dev team eg. Performance, Quality, Maintainability. This will help determine the ranking in relation to feature development

### Architecture Elaboration
Does it requires architecture elaboration?

### Tasks
- [ ] Task 1

@Klaus-Keller @devinea Please review the proposal thx.

**Comment by Klaus-Keller** [comment]
@IainSAP, why should we do this for `@sap/cds` versions < 7: 

```typescript
function updateToLegacyCdsServicePath(capServicePath: string): string {
    return uniformUrl(capServicePath).replace('odata/v4/', '');
}
```
?

Let's assume the `manifest.json` has still the `@sap/cds@6` notation, but the project has dependency to `@sap/cds@7`. With your proposal we might fix tools issues, but it would still fail when the CAP project is deployed. 
Best what we could do is to warn the developer, I don't think we should secretly 'fix' such a inconsistency for tools usage.

**Comment by devinea** [comment]
Logic such as `updateToLegacyCdsServicePath` was previously needed in Tools-suite. But tools now show better information for such errors that the user should have enough information now to correct it. 

However I think we need to allow callers `getCapModelAndServices` to skip loading the fallback to global cds version 
https://github.com/SAP/open-ux-tools/blob/f3acc1c635485fa162e9176155fd8e91f9f1c6a4/packages/project-access/src/project/cap.ts#L389-L396
@Klaus-Keller What do you think?
This would allow loading only the correct cds version from the project and using the correct URI format or throw a suitable error if `@sap/cds` is not installed in the project. If error is throw caller or user can do the `npm install` for the project and try again. 
This overlaps somewhat with https://github.com/SAP/open-ux-tools/issues/1777

**Comment by Klaus-Keller** [comment]
@devinea 
This should be addressed with PR: https://github.com/SAP/open-ux-tools/pull/1823. What is open is your [comment about CAP Java](https://github.com/SAP/open-ux-tools/pull/1823#pullrequestreview-1997363091). For CAP versions we could read the cds version from `pom.xml` and if global installed cds has different major version we throw an error. 

Should we use this ticket to address that CAP Java issue?

**Comment by IainSAP** [comment]
@Klaus-Keller Can we please update the thrown error to be of a specific error type.  There are many errors that can be thrown when compiling cds and consumers of the API need to know which specific error was thrown so the UX messaging can be accurate. https://github.com/SAP/open-ux-tools/blob/main/packages/project-access/src/project/cap.ts#L411 throws a generic error with a string, how can we tell what error was thrown unless we parse the string which is obviously not robust?

**Comment by Klaus-Keller** [comment]
> @Klaus-Keller Can we please update the thrown error to be of a specific error type. There are many errors that can be thrown when compiling cds and consumers of the API need to know which specific error was thrown so the UX messaging can be accurate. https://github.com/SAP/open-ux-tools/blob/main/packages/project-access/src/project/cap.ts#L411 throws a generic error with a string (not necessarily the string that should be presented to the user), how can we tell what error was thrown unless we parse the string which is obviously not robust? Thrown errors should also be doc'd so consumer can code appropriately. Can we add an error type definition please?

-> https://github.com/SAP/open-ux-tools/pull/1838

**Comment by devinea** [comment]
> @devinea This should be addressed with PR: #1823. What is open is your [comment about CAP Java](https://github.com/SAP/open-ux-tools/pull/1823#pullrequestreview-1997363091). For CAP versions we could read the cds version from `pom.xml` and if global installed cds has different major version we throw an error.
> 
> Should we use this ticket to address that CAP Java issue?

@Klaus-Keller  thanks for the updates. Yeah would be nice if you can also cover CAP Java in this one too.

**Comment by IainSAP** [comment]
Closing as the new error has been implemented.
